backend/src/Edna/Upload/DB/Query.hs

```diff
@@ -29,36 +32,52 @@ import Edna.Util
-insertExperiments experiments = map (fromSqlSerial . eExperimentId) .
+insertExperiments ::
+  ExperimentFileId -> [(CompoundId, TargetId)] -> Edna [ExperimentId]
+insertExperiments (SqlId experimentFileId) experiments =
+  map (fromSqlSerial . eExperimentId) .
   sortWith (\a -> (eCompoundId a, eTargetId a)) <$>
     runInsertReturningList' (insert (esExperiment ednaSchema) $ insertExpressions $
-      flip map experiments $ \(SqlId experimentFileId, SqlId compoundId, SqlId targetId) ->
+      flip map experiments $ \(SqlId compoundId, SqlId targetId) ->
         ExperimentRec
         { eExperimentId = default_
         , eExperimentFileId = val_ experimentFileId
         , eCompoundId = val_ compoundId
         , eTargetId = val_ targetId
         })
 
--- | Insert sub-experiments and return their IDs
--- Returned IDs are sorted by experiment ids of sub-experiment
-insertSubExperiments :: [Params4PLResp] -> Edna [SubExperimentId]
-insertSubExperiments params = map (fromSqlSerial . seSubExperimentId) .
-  sortWith seExperimentId <$>
-    runInsertReturningList' (insert (esSubExperiment ednaSchema) $ insertExpressions $
-      flip map params $ \Params4PLResp{plrspExperiment = SqlId experimentId, ..} ->
-        SubExperimentRec
-        { seSubExperimentId = default_
-        , seName = val_ "Primary"
-        , seAnalysisMethodId = val_ theOnlyAnalysisMethodId
-        , seExperimentId = val_ experimentId
-        , seIsSuspicious = val_ False
-        , seResult = val_ (PgJSON plrspData)
-        })
+-- | Insert sub-experiments and return their IDs.
+-- Returned records are grouped and sorted by experiment ids of sub-experiment.
+-- For each experiment one or two sub-experiments is inserted, depending on the
+-- 'plrspNewSubExp' field. Primary sub-experiment is always in the head.
+insertSubExperiments ::
+  [(ExperimentId, Either Text Params4PLResp)] -> Edna [NonEmpty SubExperimentRec]
+insertSubExperiments params =
+  -- If name is 'primaryName', the comparison will return 'False' which is less
+  -- than 'True', so primary goes first.
+  map (NE.sortWith (\se -> seName se /= primaryName)) .
+  NE.groupAllWith seExperimentId <$>
+  runInsertReturningList' (insert (esSubExperiment ednaSchema) $ insertExpressions $
+    flip foldMap params $ \(SqlId experimentId, eitherResp) ->
+      let se name result = SubExperimentRec
+            { seSubExperimentId = default_
+            , seName = val_ name
+            , seAnalysisMethodId = val_ theOnlyAnalysisMethodId
+            , seExperimentId = val_ experimentId
+            , seIsSuspicious = val_ False
+            , seResult = val_ (PgJSON result)
+            }
+      in catMaybes
+      [ Just $ se primaryName $ plrspParams <$> eitherResp
+      , (rightToMaybe eitherResp >>= plrspNewSubExp) <&>
+        \(_outliers, params4PL) -> se autoName $ Right params4PL
+      ])
+  where
+    autoName = "Automatic"
+    primaryName = "Primary"
```

Comment:
We still need to convert them to text because users may give arbitrary names to sub-experiments. So I think in most places we'll still operate with `Text` and the benefit of using ADT will be too small.